src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs

```diff
@@ -0,0 +1,233 @@
+                // don't have to worry about comments/excluded code since there are no semantic tags we want to override.
+                if (TryProcessSyntacticStringLiteral())
+                    continue;
+
+                totalTags.Add(currentSyntactic);
+                currentSyntactic = GetNextSyntacticSpan();
+            }
+
+            // We've added almost all the syntactic and semantic tags (properly skipping any semantic tags that are
+            // overridden by comments or excluded code).  All that remains is adding back the string literals we
+            // skipped.  However, when we do so, we'll see if those string literals themselves should be overridden
+            // by any embedded classifications.
+            AddEmbeddedClassifications();
+
+            return totalTags;
+
+            bool TryProcessSyntacticStringLiteral()
+            {
+                if (currentSyntactic.Tag.ClassificationType.Classification is not ClassificationTypeNames.StringLiteral and not ClassificationTypeNames.VerbatimStringLiteral)
+                    return false;
+
+                stringLiterals.Add(currentSyntactic);
+                currentSyntactic = GetNextSyntacticSpan();
+                return true;
+            }
+
+            bool TryProcessCommentOrExcludedCode()
+            {
+                if (currentSyntactic.Tag.ClassificationType.Classification is not ClassificationTypeNames.Comment and not ClassificationTypeNames.ExcludedCode)
+                    return false;
+
+                // Keep skipping semantic tags that overlaps with this syntactic tag.
+                while (currentSemantic != null && currentSemantic.Span.OverlapsWith(currentSyntactic.Span.Span))
+                    currentSemantic = GetNextSemanticSpan();
+
+                // now add that syntactic span.
+                totalTags.Add(currentSyntactic);
+                currentSyntactic = GetNextSyntacticSpan();
+                return true;
+            }
+
+            void AddEmbeddedClassifications()
+            {
+                // nothing to do if we didn't run into any string literals.
+                if (stringLiterals.Count == 0)
+                    return;
+
+                // Only need to ask for the spans that overlapped the string literals.
+                var stringLiteralSpans = new NormalizedSnapshotSpanCollection(stringLiterals.Select(s => s.Span));
+                var embeddedClassifications = embeddedTagger.GetTags(stringLiteralSpans);
+
+                // Nothing complex to do if we got no embedded classifications back.  Just add in all the string
+                // classifications, untouched.
+                if (embeddedClassifications.Count == 0)
+                {
+                    totalTags.AddRange(stringLiterals);
+                    return;
+                }
+
+                // ClassifierHelper.MergeParts requires these to be sorted.
+                stringLiterals.Sort(s_spanComparison);
```

Comment:
>stringLiterals.Sort(s_spanComparison); [](http://example.com/codeflow?start=16&length=38)

I guess being explicit is good, but isn't NormalizedSnapshotSpanCollection guaranteed ordered?

Reply:
whoops, this is over the literals not the spans, never mind